File: python/tests/conformance/support.py
"""Shared conformance helpers for Python."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable

import pytest


@dataclass
class ResolvedExpected:
    """Resolved expected outcome for a test case."""

    result: Any | None = None
    error: str | None = None
    skip: str | None = None


@dataclass
class ConformanceCase:
    """A shared conformance case."""

    id: str
    input_data: Any
    expected: ResolvedExpected


_REPO_ROOT = Path(__file__).resolve().parents[3]
_INPUT_ROOT = _REPO_ROOT / "test" / "input"
_EXPECTED_ROOT = _REPO_ROOT / "test" / "expected"
_LANGUAGE_EXCEPTION_ROOT = _REPO_ROOT / "test" / "language-exceptions"


def _read_json(path: Path) -> Any:
    return json.loads(path.read_text())


def _coerce_shared_json_int(value: Any) -> Any:
    """Convert shared JSON numeric strings into Python ints."""
    if isinstance(value, str) and value.isdigit():
        return int(value)
    return value


def _normalize_input_data(suite: str, input_data: Any) -> Any:
    """Normalize shared fixture input at the Python conformance boundary."""
    if suite == "calendar" and isinstance(input_data, dict):
        return {
            **input_data,
            "month": _coerce_shared_json_int(input_data["month"]),
            "year": _coerce_shared_json_int(input_data["year"]),
        }

    if suite == "vbml" and isinstance(input_data, dict):
        components = []
        for component in input_data.get("components", []):
            if not isinstance(component, dict) or "calendar" not in component:
                components.append(component)
                continue

            calendar = component.get("calendar")
            if not isinstance(calendar, dict):
                components.append(component)
                continue

            components.append(
                {
                    **component,
                    "calendar": {
                        **calendar,
                        "month": _coerce_shared_json_int(calendar["month"]),
                        "year": _coerce_shared_json_int(calendar["year"]),
                    },
                }
            )

        return {**input_data, "components": components}

    return input_data


def _walk_json_files(root: Path) -> list[Path]:
    return sorted(path for path in root.rglob("*.json") if path.is_file())


def _resolve_expected(
    suite: str,
    case_id: str,
    payload: dict[str, Any],
    language: str,
) -> ResolvedExpected:
    exception_path = _LANGUAGE_EXCEPTION_ROOT / language / suite / f"{case_id}.json"

    if exception_path.exists():
        exception = _read_json(exception_path)
        reason = str(exception.get("reason", "")).strip()
        if not reason:
            raise ValueError(
                f'Language exception "{suite}/{case_id}" is missing a reason.'
            )

        if exception.get("skip"):
            if "result" in exception or "error" in exception:
                raise ValueError(
                    f'Language exception "{suite}/{case_id}" cannot define '
                    "skip with result or error."
                )

            return ResolvedExpected(skip=reason)

        has_result = "result" in exception
        has_error = "error" in exception
        if has_result == has_error:
            raise ValueError(
                f'Language exception "{suite}/{case_id}" must define exactly '
                "one of result or error."
            )

        return ResolvedExpected(
            result=exception.get("result"),
            error=exception.get("error", {}).get("message"),
        )

    has_result = "result" in payload
    has_error = "error" in payload
    if has_result == has_error:
        raise ValueError(
            f'Conformance case "{suite}/{case_id}" must define exactly one of '
            "result or error."
        )

    return ResolvedExpected(
        result=payload.get("result"),
        error=payload.get("error", {}).get("message"),
    )


def load_cases(suite: str, language: str = "python") -> list[ConformanceCase]:
    """Load all cases for a suite."""

    input_root = _INPUT_ROOT / suite
    expected_root = _EXPECTED_ROOT / suite

    cases: list[ConformanceCase] = []
    for input_path in _walk_json_files(input_root):
        relative_path = input_path.relative_to(input_root)
        expected_path = expected_root / relative_path

        if not expected_path.exists():
            raise FileNotFoundError(f"Missing expected file for {suite}/{relative_path}")

        case_id = str(relative_path.with_suffix("")).replace("\\", "/")
        input_data = _normalize_input_data(suite, _read_json(input_path))
        expected_payload = _read_json(expected_path)

        cases.append(
            ConformanceCase(
                id=case_id,
                input_data=input_data,
                expected=_resolve_expected(suite, case_id, expected_payload, language),
            )
        )

    return cases


def case_ids(case: ConformanceCase) -> str:
    """Stable pytest case ids."""

    return case.id


def assert_case(case: ConformanceCase, run: Callable[[Any], Any]) -> None:
    """Run and assert a shared conformance case."""

    if case.expected.skip:
        pytest.skip(case.expected.skip)

    if case.expected.error:
        with pytest.raises(Exception, match=re.escape(case.expected.error)):
            run(case.input_data)
        return

    assert run(case.input_data) == case.expected.result

File: src/__tests__/conformance/support.ts
import fs from "fs";
import path from "path";

interface ConformanceErrorExpectation {
  message: string;
}

interface PlatformException<TExpected> {
  reason: string;
  result?: TExpected;
  error?: ConformanceErrorExpectation;
  skip?: boolean;
}

interface ConformanceExpected<TExpected> {
  result?: TExpected;
  error?: ConformanceErrorExpectation;
}

interface ConformanceCase<TInput, TExpected> {
  id: string;
  input: TInput;
  expected: ConformanceExpected<TExpected>;
}

interface ResolvedExpectation<TExpected> {
  result?: TExpected;
  error?: ConformanceErrorExpectation;
  skip?: string;
}

interface RunConformanceSuiteOptions<TInput, TExpected> {
  suiteName: string;
  suiteDir: string;
  run: (input: TInput) => TExpected;
  platform?: string;
}

const readJsonFile = <T>(filePath: string): T =>
  JSON.parse(fs.readFileSync(filePath, "utf8")) as T;

const walkJsonFiles = (rootDir: string): string[] => {
  const entries = fs.readdirSync(rootDir, {
    withFileTypes: true,
  });

  return entries
    .flatMap((entry) => {
      const fullPath = path.join(rootDir, entry.name);

      if (entry.isDirectory()) {
        return walkJsonFiles(fullPath);
      }

      return entry.name.endsWith(".json") ? [fullPath] : [];
    })
    .sort();
};

const loadCases = <TInput, TExpected>(suiteDir: string): Array<ConformanceCase<TInput, TExpected>> => {
  const inputRoot = path.resolve(__dirname, "../../../test/input", suiteDir);
  const expectedRoot = path.resolve(__dirname, "../../../test/expected", suiteDir);

  return walkJsonFiles(inputRoot).map((inputPath) => {
    const relativePath = path.relative(inputRoot, inputPath);
    const expectedPath = path.join(expectedRoot, relativePath);

    if (!fs.existsSync(expectedPath)) {
      throw new Error(
        `Missing expected file for ${suiteDir}/${relativePath}.`
      );
    }

    return {
      id: relativePath.replace(/\.json$/, ""),
      input: readJsonFile<TInput>(inputPath),
      expected: readJsonFile<ConformanceExpected<TExpected>>(expectedPath),
    };
  });
};

const loadPlatformException = <TExpected>(
  platform: string,
  suiteDir: string,
  testId: string
): PlatformException<TExpected> | null => {
  const exceptionPath = path.resolve(
    __dirname,
    "../../../test/platform-exceptions",
    platform,
    suiteDir,
    `${testId}.json`
  );

  if (!fs.existsSync(exceptionPath)) {
    return null;
  }

  return readJsonFile<PlatformException<TExpected>>(exceptionPath);
};

const resolveExpectation = <TExpected>(
  platform: string,
  suiteDir: string,
  testCase: ConformanceCase<unknown, TExpected>
): ResolvedExpectation<TExpected> => {
  const exception = loadPlatformException<TExpected>(
    platform,
    suiteDir,
    testCase.id
  );

  if (exception) {
    if (!exception.reason.trim()) {
      throw new Error(
        `Platform exception "${suiteDir}/${testCase.id}" is missing a reason for ${platform}.`
      );
    }

    if (exception.skip) {
      if (exception.result !== undefined || exception.error !== undefined) {
        throw new Error(
          `Platform exception "${suiteDir}/${testCase.id}" cannot define skip with result or error.`
        );
      }

      return {
        skip: exception.reason,
      };
    }

    if ((exception.result === undefined) === (exception.error === undefined)) {
      throw new Error(
        `Platform exception "${suiteDir}/${testCase.id}" must define exactly one of result or error.`
      );
    }

    return {
      result: exception.result,
      error: exception.error,
    };
  }

  if ((testCase.expected.result === undefined) === (testCase.expected.error === undefined)) {
    throw new Error(
      `Conformance case "${suiteDir}/${testCase.id}" must define exactly one of result or error.`
    );
  }

  return {
    result: testCase.expected.result,
    error: testCase.expected.error,
  };
};

export const runConformanceSuite = <TInput, TExpected>({
  suiteName,
  suiteDir,
  run,
  platform = "node",
}: RunConformanceSuiteOptions<TInput, TExpected>): void => {
  const cases = loadCases<TInput, TExpected>(suiteDir);

  describe(suiteName, () => {
    cases.forEach((testCase) => {
      const resolved = resolveExpectation(platform, suiteDir, testCase);
      const testMethod = resolved.skip ? it.skip : it;
      const label = testCase.id;

      testMethod(label, () => {
        if (resolved.error) {
          expect(() => run(testCase.input)).toThrow(resolved.error.message);
          return;
        }

        expect(run(testCase.input)).toEqual(resolved.result);
      });
    });
  });
};
